flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBIncrementalRestoreOperation.java

```diff
@@ -329,74 +332,164 @@ private void restoreWithRescaling(Collection<KeyedStateHandle> restoreStateHandl
+                        tmpRestoreDBInfo.columnFamilyHandles;
+
+                // iterating only the requested descriptors automatically skips the default
+                // column
+                // family handle
+                for (int descIdx = 0; descIdx < tmpColumnFamilyDescriptors.size(); ++descIdx) {
+                    ColumnFamilyHandle tmpColumnFamilyHandle = tmpColumnFamilyHandles.get(descIdx);
+
+                    ColumnFamilyHandle targetColumnFamilyHandle =
+                            this.rocksHandle.getOrRegisterStateColumnFamilyHandle(
+                                            null,
+                                            tmpRestoreDBInfo.stateMetaInfoSnapshots.get(descIdx))
+                                    .columnFamilyHandle;
+
+                    try (RocksIteratorWrapper iterator =
+                            RocksDBOperationUtils.getRocksIterator(
+                                    tmpRestoreDBInfo.db,
+                                    tmpColumnFamilyHandle,
+                                    tmpRestoreDBInfo.readOptions)) {
+
+                        iterator.seek(startKeyGroupPrefixBytes);
+
+                        while (iterator.isValid()) {
+
+                            if (RocksDBIncrementalCheckpointUtils.beforeThePrefixBytes(
+                                    iterator.key(), stopKeyGroupPrefixBytes)) {
+                                writeBatchWrapper.put(
+                                        targetColumnFamilyHandle, iterator.key(), iterator.value());
+                            } else {
+                                // Since the iterator will visit the record according to the
+                                // sorted
+                                // order,
+                                // we can just break here.
+                                break;
+                            }
+
+                            iterator.next();
+                        }
+                    } // releases native iterator resources
+                }
+                logger.info(
+                        "Finished restoring from state handle: {} with rescaling.", stateHandle);
+            }
+        }
+    }
+
+    /**
+     * Recovery from multi incremental states with rescaling. For rescaling, this method creates a
+     * temporary RocksDB instance for a key-groups shard. All contents from the temporary instance
+     * are copied into the real restore instance and then the temporary instance is discarded.
+     */
+    private void rescaleClipIngestDB(
+            Collection<IncrementalLocalKeyedStateHandle> localKeyedStateHandles,
+            byte[] startKeyGroupPrefixBytes,
+            byte[] stopKeyGroupPrefixBytes)
+            throws Exception {
+
+        final Path absolutInstanceBasePath = instanceBasePath.getAbsoluteFile().toPath();
+        final Path exportCfBasePath = absolutInstanceBasePath.resolve("export-cfs");
```

Comment:
Just to clarify, this is not a temporary folder so that the files can be efficiently sym-linked?